src/org/rascalmpl/dap/RascalDebugAdapter.java

```diff
@@ -506,48 +521,221 @@ public CompletableFuture<Void> pause(PauseArguments args) {
+                        completion.setText(funcName.substring(text.length()));
+                    }
+                    else { // exact match
+                        completion.setText("");
+                    }
+                    completion.setType(CompletionItemType.FUNCTION);
+                    completion.setDetail(getFunctionDetail(func));
+                    completions.add(completion);
+                }
+            }
+        }
+    }
+
+    @Override
+    public CompletableFuture<CompletionsResponse> completions(CompletionsArguments args) {
+        return CompletableFuture.supplyAsync(() -> {
+            CompletionsResponse response = new CompletionsResponse();
+            List<CompletionItem> completions = new ArrayList<>();
+            if(suspendedState.isSuspended()) {
+                Integer frameId = args.getFrameId();
+                assert frameId != null;
+                IRascalFrame frame = suspendedState.getCurrentStackFrames()[frameId];
+
+                // Variable names starting with the typed text
+                addVariablesCompletionsFromFrame(frame, args.getText(), completions);
+
+                // Add functions from current frame's module
+                if(frame instanceof Environment){
+                    addFunctionsCompletionsFromEnvironment((Environment) frame, args.getText(), completions);
+                }
+
+                // Add module for namespace calls
+                for(String importName : frame.getImports()){
+                    if(importName.startsWith(args.getText())){
+                        CompletionItem completion = new CompletionItem();
+                        completion.setLabel(importName);
+                        completion.setSortText(importName);
+                        if (args.getText().length() < importName.length()) { //remove the prefix
+                            completion.setText(importName.substring(args.getText().length()) + "::");
+                        }
+                        else { // exact match
+                            completion.setText("::");
+                        }
+                        completion.setType(CompletionItemType.MODULE);
+                        completions.add(completion);
+                    }
+                }
+
+                // Add functions from imported modules
+                // First check for prefix with ::
+                String[] parts = args.getText().split("::");
+                Set<String> importsToSearch = parts.length > 1 ? Set.of(parts[0]) : frame.getImports();
+                String functionToSearch = parts.length > 1 ? parts[1] : args.getText();
+                for(String importName : importsToSearch){
+                    IRascalFrame module = evaluator.getModule(importName);
+                    // Try to cast module IRascalFrame as an Environment to get its functions
+                    if(module != null && module instanceof Environment){
```

Comment:
nit: `module != null` can be skipped, as `instanceof` always returns `false` for `null`.